### ISSUE TITLE
Improved handling of AuthenticationFailedException in ApiResponseHelper.RunCommandAsync

### ISSUE THREAD

**Issue by JornWildt** [cofoundry-cms/cofoundry#415]
It is very neat to be able to write API handlers like this and know the framework will do validation and other checks:

```
    [HttpPost]
    public async Task<JsonResult> Post([FromQuery] int id, [FromBody] MyCommand command)
    {
      command.Id = id;
      return await ApiResponseHelper.RunCommandAsync(command);
    }
```

But if my command handler throws `AuthenticationFailedException` then the above API code fails with an HTTP status code 500 "Internal error" and returns a stacktrace in HTML - which is no good for error handling in JavaScript.

Please handle `AuthenticationFailedException` in `ApiResponseHelper.RunCommandAsync` and return an appropriate HTTP status code 403 "Forbidden".

The same applies to `EntityNotFoundException` which I would expect to return HTTP status code 404 and a JSON response.

Generally speaking, all exceptions should be handled by `ApiResponseHelper.RunCommandAsync` and return a suitable JSON representation of the error - including the stacktrace in a property - and may very well also use HTTP status code 500.

**Comment by HeyJoel** [comment]
So, we do handle `NotPermittedException` as a 403 as you expect, but not `AuthenticationFailedException`, which I can't actually see being used anywhere internally by Cofoundry so I might need to look further into that, it could be that it's been superseded by more specific exceptions like `InvalidCredentialsAuthenticationException` for logging in, and was never removed. Those exceptions inherit from ValidationException and will therefore return a 400. 

Did you see `AuthenticationFailedException` in an example somewhere? I can only think that it would be used for logging in, or perhaps authenticating an API token. If it's useful I can certainly add it into the `ApiResponseHelper`

For `EntityNotFoundException`, this is designed to be used when an entity is required/expected but could not be found, and should therefore I think should be an error 500.  However, this exception does get used in two different situations, for example when updating an entity:

1. If the entity to update does not exist, then this error is thrown
2. If a related entity that we expect exists does not exist, then this error is thrown.

I suppose in terms of API design, in the first example I would expect a 404, but in the second example I think a 404 is misleading - the entity does exist, but some related data could not be found,  and that should be exceptional. I suppose we could throw two different exception types here and handle them differently, although it feels more about being semantically correct than being useful, but I am certainly open to that.

For "get" operations we shouldn't be throwing `EntityNotFoundException`, and returning `null` will result in a 404 for singular results. Collections will just return the empty collection.

**Comment by JornWildt** [comment]
> So, we do handle NotPermittedException as a 403 as you expect, but not AuthenticationFailedException

Okay, I'll change my code, thanks. `AuthenticationFailedException` was just the first exception on the "find *exception" list.

> For "get" operations we shouldn't be throwing EntityNotFoundException, and returning null will result in a 404 for singular results. 

Fine, that will work too, I'll change my code. Thanks 👍

It is mostly a matter of style - and I'm coming from a system where we throw `MissingResourceException` whenever something is not found - so throwing `EntityNotFoundException` was a natural choice :-)

Maybe a chapter on API error handling could be added to the backlog?

**Comment by JornWildt** [comment]
I think the same applies to PermissionValidationFailedException which I get in this code:

```
[HttpPost]
public async Task<JsonResult> Get([FromBody] SearchBookingSummariesQuery query)
{
  IList<BookingSummary> bookings = await QueryExecutor.ExecuteAsync(query);
  return ApiResponseHelper.SimpleQueryResponse(bookings);
}
```

When the query handler fails permission checks it throws `PermissionValidationFailedException` which in turn returns an HTML stack trace and HTTP status code 500.

I would have expected HTTP status code 403 and a JSON response.

The query handler implements `IPermissionRestrictedQueryHandler<SearchBookingSummariesQuery, IList<BookingSummary>>` and as such it leaves the permission checking to Cofoundry.

**Comment by HeyJoel** [comment]
Hey sorry, just getting on to this now...

So the reason you get the exception from the query is that the query isn't being run by `IApiResponseHelper`, I think originally I didn't expect queries to need error handling so I didn't implement "run" methods for them in the helper, but obviously they do for some edge cases. 

Generally I need to look at `ApiResponseHelper` and working with APIs. The helper is still relevant for Cofoundry internal APIs where we need to be unaffected by customizations a developer may make to the underlaying ASP.NET pipeline, but for public use I think it would be better to work with what ASP.NET is doing and provide some light helper functionality around that. Ideally we just want to provide the data and for you to format it however you want rather than provide a very opinionated way to do APIs.

Anyway, that's a job for another day, for now I've added some query helper methods to `IApiResponseHelper` so you can now do things like:

```csharp
[HttpPost]
public async Task<JsonResult> Get(SearchBookingSummariesQuery query)
{
    return await _apiResponseHelper.RunQueryAsync(query);
}
```

Or 

```csharp
public async Task<JsonResult> GetById(int pageId)
{
    return await _apiResponseHelper.RunWithResultAsync(() => _contentRepository
        .Pages()
        .GetById(pageId)
        .AsDetails()
        .ExecuteAsync()
        );
}
```

This change will be released in v0.10
